### ISSUE TITLE
Send Later silently fails to deliver messages > mailnews.message_warning_size

### ISSUE THREAD

**Issue by jikamens** [Extended-Thunder/send-later#643]
### Prerequisites

- [X] I am reporting a problem, not asking a question (for questions, go to [Discussions](https://github.com/Extended-Thunder/send-later/discussions).
- [X] I have consulted the [user guide](https://extended-thunder.github.io/send-later/).
- [X] I have confirmed that I'm running the current version and the problem persists there (note the version number in fine print next to the add-on name [on addons.thunderbird.net](https://addons.thunderbird.net/thunderbird/addon/send-later-3/) and compare it to the version number listed in the add-on details opened from the Add-ons page in Thunderbird; if you've got an old version, click the gear icon and select "Check for Updates").
- [X] If I'm having an issue with date-parsing confusion, I've read the [section about that in the user guide](https://extended-thunder.github.io/send-later/#date-format-confusion-in-the-scheduler-pop-up) and I still think this is something different.
- [X] If I'm having an issue with messages not being delivered, I've followed the [troubleshooting steps in the user guide](https://extended-thunder.github.io/send-later/#messages-dont-send-or-send-multiple-times) and it's still an issue.

### Describe the bug

As described in https://bugzilla.mozilla.org/show_bug.cgi?id=1853253, messages sent through the Outbox when Send Later isn't installed don't get a large-message warning from Thunderbird but then can't be sent; Thunderbird claims it's sending them but gets stuck there forever.

Until this is fixed, Send Later needs to detect if a message is too large and prevent the user from scheduling it if so.

### Thunderbird version

115

**Comment by jikamens** [comment]
Hmm. I started working on adding a warning to Send Later, and I have concerns. Specifically, it's impossible for Send Later to find out what the size of the saved draft is until after it's saved with the Send Later schedule. So to warn and not schedule the message at that point would mean one of two approaches (all of this assumes we've fetched the size limit preference and it's greater than 0):

1. Save the draft without the Send Later headers.
2. Read it back to check the size.
3. Alert and refuse to schedule the message if it's close enough to the message limit, like within 1% to give us a margin of error.
4. Otherwise save the draft again with the Send Later headers.

This has significant performance implications, especially if the user's Drafts folder isn't synchronized locally, because we're adding an extra message read and extra message write, which will be especially noticeable for large messages.

Option 2:

1. Save the draft with the Send Later headers.
2. Read it back to check the size.
3. If it's too large, save the message again, this time without the headers, and alert the user.

This adds an extra message read but only adds an extra message write if the message turns out to be too large. However, it has another issue: it introduces a race condition, since in between step 1 and step 3 the Send Later main loop could decide to send the message. I suppose we could introduce some sort of locking to prevent that from happening? But what about if the user is doing server-side send later so the message is actually being delivered on a different computer? Ugh that won't be reliable either.

It seems like the only reliable way to do this that doesn't introduce a significant performance impact is to only do the check at delivery time and refuse to copy the message into the Outbox if it's too large. That's certainly better than what we've got now, but the UX isn't great because up until the delivery time the user will think the message is going to be sent successfully. Still, I suppose it's worth doing just to prevent Thunderbird from getting totally wedged on a too-large message in the Outbox.

**Comment by jikamens** [comment]
The workaround I just committed to main is better than nothing but it's not the real fix. The real fix is blocked on a Thunderbird fix (see bug linked above).
